### PR TITLE
Add ScalarValue support for arbitrary list elements

### PR DESCRIPTION
**Pull request by jonmmease**
# Which issue does this PR close?
Closes https://github.com/apache/arrow-datafusion/issues/1092

 # Rationale for this change

See https://github.com/apache/arrow-datafusion/issues/1092.

# What changes are included in this PR?

The `ScalarValue::iter_to_array` and `ScalarValue::to_array_of_size` associated functions have been updated to support `ScalarValue::List` variants that contain arbitrary `ScalarValue` elements.  Previously, only lists of primitive elements were supported.

# Are there any user-facing changes?

Previously, an error was returned when calling `iter_to_array` or `to_array_of_size` with lists containing non-primitive values. Now these calls complete successfully.

No

### CODE REVIEW

File: datafusion/src/scalar.rs [hunk 1]
```diff
@@ -833,6 +838,74 @@ impl ScalarValue {
...
+
+        let mut elements: Vec<ArrayRef> = Vec::new();
+        let mut valid = BooleanBufferBuilder::new(0);
+        let mut flat_len = 0i32;
+        for scalar in scalars {
+            if let ScalarValue::List(values, _) = scalar {
+                match values {
+                    Some(values) => {
+                        let element_array =
+                            ScalarValue::iter_to_array(values.as_ref().clone())?;
```
Comment (alamb) [inline datafusion/src/scalar.rs:858]:
```suggestion
                        let element_array =
                            ScalarValue::iter_to_array(*value)?;
```

I think you can avoid the clone by unboxing the values (I definitely get confused that a `*` deref of a box is how one unboxes values)

File: datafusion/src/scalar.rs [hunk 2]
```diff
@@ -833,6 +838,74 @@ impl ScalarValue {
...
+
+        // Concatenate element arrays to create single flat array
+        let element_arrays: Vec<&dyn Array> =
+            elements.iter().map(|a| a.as_ref()).collect();
+        let flat_array = match arrow::compute::concat(&element_arrays) {
+            Ok(flat_array) => flat_array,
+            Err(err) => return Err(DataFusionError::ArrowError(err)),
+        };
+
+        // Build ListArray using ArrayData so we can specify a flat inner array, and offset indices
```
Comment (alamb) [inline datafusion/src/scalar.rs:897]: Looks good 👍

File: datafusion/src/scalar.rs [hunk 3]
```diff
@@ -833,6 +838,74 @@ impl ScalarValue {
...
+        let offsets_array = offsets.finish();
+        let array_data = ArrayDataBuilder::new(data_type.clone())
+            .len(offsets_array.len() - 1)
+            .null_bit_buffer(valid.finish())
+            .add_buffer(offsets_array.data().buffers()[0].clone())
+            .add_child_data(flat_array.data().clone());
+
+        let list_array = ListArray::from(array_data.build());
```
Comment (alamb) [inline datafusion/src/scalar.rs:905]: When you rebase this PR, you will likely have to change this to `array_data.build()?` as `build()` is fallible as of arrow-6.0.0

File: datafusion/src/scalar.rs [hunk 4]
```diff
@@ -945,7 +1018,15 @@ impl ScalarValue {
                 &DataType::LargeUtf8 => {
                     build_list!(LargeStringBuilder, LargeUtf8, values, size)
                 }
-                dt => panic!("Unexpected DataType for list {:?}", dt),
+                _ => ScalarValue::iter_to_array_list(
```
Comment (alamb) [inline datafusion/src/scalar.rs:1021]: I wonder if we could remove all the other `DataType` matches above and consolidate the code to use `iter_to_array_list` rather than `build_list!`

Though perhaps that code cleanup would lose some optimizations (aka would concat a lot of small arrays) 🤔

**Review by alamb — APPROVED** [review]
Thank you for the contribution. looks good to me @jonmmease  -- can you please rebase this PR against master?

File: datafusion/src/scalar.rs (hunk 4 above)
Comment (jonmmease) [reply datafusion/src/scalar.rs:1020]: My intuition was that the current specialized implementations would be more efficient, but I didn't actually benchmark that assumption.

File: datafusion/src/scalar.rs (hunk 3 above)
Comment (jonmmease) [reply datafusion/src/scalar.rs]: Done in 1a94255

File: datafusion/src/scalar.rs (hunk 1 above)
Comment (jonmmease) [reply datafusion/src/scalar.rs]: Ahh, thanks. Done in 1a94255

**Comment by alamb** [comment]
> Out of curiosity, when you say "rebase against master", do you mean git rebase or git merge? I just wanted to clarify since I know git rebase clobbers the commit history and can make things harder to review.

TLDR you are right that `git merge` might be a better suggestion

I think I was using `git rebase` because of some long  time again when a `git merge` in the github ui showed all commits that were merged in even if they weren't part of the PR. 

Since we squash into a single commit on merge anyway, I don't think it makes any difference to the final result

Sorry for any confusion

File: datafusion/src/scalar.rs (hunk 4 above)
Comment (alamb) [reply datafusion/src/scalar.rs:1020]: Leaving it this way is probably the best plan

**Review by alamb — APPROVED** [review]
